_posts/2020-09-22-rust-2021.md

```diff
@@ -0,0 +1,96 @@
+
+## Baby steps
+
+While I am driven by a long-term, ambitious vision, the goal of Druid in 2021 is not to deliver a general UI toolkit. Rather, we are deliberately continuing to follow a narrow scope. The primary goal remains the font editor project, and we plan to re-focus attention on that. I do think this is an attainable goal. I also think that what we learn from trying to build a real application with users will be extremely valuable to the more ambitious task.
+
+One project management technique that is proving effective is "cycles." Instead of trying to solve the most ambitious version of a problem, we choose up front what to push to a future cycle, reducing the scope for the current implementation cycle. An example is the choice to defer BiDi from our recent text work. This is obviously an essential feature for a real GUI toolkit, but we also know it could take weeks or months to get it right. To have any chance of shipping, we have to carefully budget our time and energy on subprojects that easily could expand to absorb our full attention.
+
+## A wishlist
+
+Here are some things I'd especially like from the rest of the Rust community.
+
+First, I'd love for wgpu to become more mature, as I think it is the most promising approach to GPU for the future. Our [roadmap for 3D] basically depends on it. Much love to that community for the progress they've made so far and hope for the future.
+
+Second, while we generally find Windows platform bindings to be straightforward, thanks to the excellent winapi bindings (and hopefully a stable [com-rs] shortly), the situation on macOS is not as good. There are real problems interoperating with Objective-C code, including type-safety in wrappers that also allow subclassing, and continual friction around autorelease pools. There are also different, incompatible conventions around wrapping foreign types (foreign_object vs TCFType). None of this is blocking, but it makes developing macOS platform integration less fun. (See [this HN thread](https://news.ycombinator.com/item?id=24309565) for a bit more background)
+
+Third, some perhaps surprisingly good news. I prefer Rust crates when possible rather than relying on C++, but for OpenType font shaping, we would have to call into HarfBuzz. Now not one but two Rust solutions are on the horizon: [Allsorts] from YesLogic, and [rustybuzz] from RazrFalcon. Both are young, but I can see adopting them (primarily for use on Linux, which is still work in progress; on Windows and macOS we use the platform text layout capabilities). I know of no other language besides C++ that has working OpenType font shaping. Almost every language community has its own language-native GUI toolkit project, and most fall short. I do believe that Rust will be different, and seeing projects like this is to me evidence why.
```

Comment:
Also agree! Rather than converging on a single vision, I believe building great crates that tackle direct use cases of a GUI toolkit should be one of our main focuses.

For instance, I took a lot of inspiration from `druid` when [I rewrote the layout engine in `iced`](https://github.com/hecrj/iced/pull/52). If we converged our visions here, we could maybe end up with a nice crate that could benefit other parts of the ecosystem. For example, [Bevy is working on their UI implementation](https://github.com/bevyengine/bevy/issues/254) and they may consider adopting a similar layout engine.

I think there is a lot of space for great crates here: layout, text caching, accessibility, windowing, etc. Collaboration is hard, however!

Reply:
I also agree. I've been focused on other areas so we haven't moved forward much in the UI department yet, but I would love to collaborate on as many areas as we can here. I'm not a UI toolkit development expert (nor do I have the bandwidth to be), so it makes way more sense to me to lend my project's development energy to ecosystem crates with a lot of brainpower and experience behind them. Ideally these crates would be modular in a way that lets us bring our own windowing and renderering.

As a small tanget, I really _want_ to discover good ways to do reactive UI in Bevy ECS, but its possible that every solution we come up with is significantly more cumbersome than just fully adopting a druid/crochet/iced/elm-like dataflow. If that ends up being the case, it would be great to be able to just pull in the relevant ecosystem crates instead of re-inventing the wheel.